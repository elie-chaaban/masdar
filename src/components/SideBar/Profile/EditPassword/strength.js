const hasNumbers = (value) => {
  return new RegExp(/[0-9]/).test(value);
};
const hasSpecial = (value) => {
  return /[!@#$^&*)(+=._-]/.test(value);
};
const hasMixed = (value) => {
  return new RegExp(/[a-z]/).test(value) && new RegExp(/[A-Z]/).test(value);
};
export const strengthColor = (count) => {
  if (count < 3) return {label: 'Very Weak', class: 'verylow'};
  if (count < 4) return {label: 'Weak', class: 'low'};
  if (count < 5) return {label: 'Strong', class: 'strong'};
  if (count < 6) return {label: 'Very Strong', class: 'verystrong'};
};
export const strengthIndicator = (value) => {
  const matched = [];
  if (value.length > 5) matched.push('greater-than-5');
  if (value.length > 7) matched.push('greater-than-7');
  if (hasNumbers(value)) matched.push('has-numbers');
  if (hasMixed(value)) matched.push('has-mixed');
  if (hasSpecial(value)) matched.push('has-special');
  return matched.length;
};
