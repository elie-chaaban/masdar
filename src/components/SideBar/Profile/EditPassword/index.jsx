import React, {useState} from 'react';
import {Formik, Form, Field} from 'formik';
import passwordValidation from '../validations/password';
import {Col, Row} from 'react-bootstrap';
import {strengthColor, strengthIndicator} from './strength';
import {updatePassword} from '../../../../services/user';
import PropTypes from 'prop-types';
import {errorNotification, successNotification} from '../../../Utils/Notifications';

const EditPassword = ({click, profile}) => {
  const [indicator, setIndicator] = useState(undefined);
  const [loading, setLoading] = useState(false);
  const checkPassStrength = (value) => {
    const count = strengthIndicator(value);
    setIndicator(strengthColor(count));
  };
  const updatePass = async (password) => {
    try {
      setLoading(true);
      await updatePassword(profile, password);
      click();
      successNotification('Password Updated Successfully!');
    } catch (error) {
      errorNotification(error);
    } finally {
      setLoading(false);
    }
  };
  return (
    <Formik
      initialValues={{
        password: '',
        passwordConfirmation: ''
      }}
      validationSchema={passwordValidation}
      onSubmit={({password}) => {
        if (!loading) updatePass(password);
      }}
    >
      {({errors, touched, handleChange}) => {
        return (
          <Form className="password-form row">
            <div className="mt-4 col-6">
              <p>Password</p>
              <Field
                className="form-control"
                placeholder="password"
                name="password"
                type="password"
                onChange={(e) => {
                  handleChange(e);
                  checkPassStrength(e.target.value);
                }}
              />
              {errors.password && touched.password ? <div className="text-danger">{errors.password}</div> : null}
            </div>
            <div className=" mt-4  col-6 d-flex flex-column justify-content-center">
              <div className="password-strength">
                <p>Password strength</p>
                <div data-testid="indicator-bar" className={`bar ${indicator ? indicator.class : ''}`}></div>
                {indicator && <p data-testid="indicator-text">{indicator.label}</p>}
              </div>
            </div>
            <div className="mt-4 col-6">
              <p>Confirm Password</p>
              <Field className="form-control" placeholder="password Confirmation" name="passwordConfirmation" type="password" />
              {errors.passwordConfirmation && touched.passwordConfirmation ? (
                <div className="text-danger">{errors.passwordConfirmation}</div>
              ) : null}
            </div>
            <Col xs={12}>
              <Row>
                <Col xs={6}>
                  <button className="btn mt-3" type="submit">
                    {loading ? 'Saving...' : 'Save'}
                  </button>
                </Col>
              </Row>
            </Col>
          </Form>
        );
      }}
    </Formik>
  );
};
EditPassword.protoTypes = {
  click: PropTypes.func.isRequired
};
export default EditPassword;
