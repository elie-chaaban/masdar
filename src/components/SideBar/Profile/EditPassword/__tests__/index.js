import React from 'react';
import EditPassword from '..';
import {act} from 'react-dom/test-utils';
import {mount} from 'enzyme';

let resolve;
jest.mock('../../../../../services/user', () => ({
  ...jest.requireActual('../../../../../services/user'),
  updatePassword: jest.fn(
    () =>
      new Promise((res) => {
        resolve = res;
      })
  )
}));

describe('Edit Password Component', () => {
  let dom;
  const click = jest.fn();
  beforeEach(() => {
    dom = mount(<EditPassword click={click} />);
  });
  afterEach(() => {
    jest.clearAllMocks();
    dom.unmount();
  });
  it('should match snapshot', () => {
    expect(dom).toMatchSnapshot();
  });
  test('default inputs', () => {
    expect(dom.find('input[name="password"]').prop('value')).toBe('');
    expect(dom.find('input[name="passwordConfirmation"]').prop('value')).toBe('');
    expect(dom.find('.text-danger')).toHaveLength(0);
    expect(dom.find('[data-testid="indicator-text"]')).toHaveLength(0);
    expect(dom.find('[data-testid="indicator-bar"]').prop('className')).toBe('bar ');
  });
  it('should validate required inputs', async () => {
    await act(async () => {
      dom.find('button').simulate('submit');
    });
    dom.update();
    expect(dom.find('.text-danger')).toHaveLength(2);
    expect(dom.find('.text-danger').at(0).text()).toBe('required');
    expect(dom.find('.text-danger').at(1).text()).toBe('required');
  });
  it('should validate password confirmation to match password', async () => {
    await act(async () => {
      await dom.find('input[name="password"]').simulate('change', {target: {value: 'pass123', name: 'password'}});
      await dom.find('input[name="passwordConfirmation"]').simulate('change', {target: {value: 'pass13', name: 'passwordConfirmation'}});
      dom.find('button').simulate('submit');
    });
    dom.update();
    expect(dom.find('.text-danger')).toHaveLength(1);
    expect(dom.find('.text-danger').text()).toBe('Passwords must match');
  });
  it('should show password indicator', async () => {
    await act(async () => {
      await dom.find('input[name="password"]').simulate('change', {target: {value: 'he', name: 'password'}});
      dom.update();
      expect(dom.find('[data-testid="indicator-bar"]').hasClass('verylow')).toBeTruthy();
      expect(dom.find('[data-testid="indicator-text"]').find('p').text()).toBe('Very Weak');
      await dom.find('input[name="password"]').simulate('change', {target: {value: 'password123', name: 'password'}});
      dom.update();
      expect(dom.find('[data-testid="indicator-bar"]').hasClass('low')).toBeTruthy();
      expect(dom.find('[data-testid="indicator-text"]').find('p').text()).toBe('Weak');
      await dom.find('input[name="password"]').simulate('change', {target: {value: 'password123@', name: 'password'}});
      dom.update();
      expect(dom.find('[data-testid="indicator-bar"]').hasClass('strong')).toBeTruthy();
      expect(dom.find('[data-testid="indicator-text"]').find('p').text()).toBe('Strong');
      await dom.find('input[name="password"]').simulate('change', {target: {value: 'Password123@', name: 'password'}});
      dom.update();
      expect(dom.find('[data-testid="indicator-bar"]').hasClass('verystrong')).toBeTruthy();
      expect(dom.find('[data-testid="indicator-text"]').find('p').text()).toBe('Very Strong');
    });
  });
  test('password submission', async () => {
    await act(async () => {
      await dom.find('input[name="password"]').simulate('change', {target: {value: 'Password123@', name: 'password'}});
      await dom
        .find('input[name="passwordConfirmation"]')
        .simulate('change', {target: {value: 'Password123@', name: 'passwordConfirmation'}});
      dom.find('button').simulate('submit');
    });
    dom.update();
    expect(dom.find('button').text()).toBe('Saving...');
    expect(click).toHaveBeenCalledTimes(0);
    await act(async () => {
      resolve();
    });
    expect(dom.find('button').text()).toBe('Save');
    expect(click).toHaveBeenCalledTimes(1);
  });
});
