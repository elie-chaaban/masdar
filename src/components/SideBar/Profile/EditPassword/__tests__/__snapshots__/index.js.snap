// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Password Component should match snapshot 1`] = `
<EditPassword
  click={[MockFunction]}
>
  <Formik
    initialValues={
      Object {
        "password": "",
        "passwordConfirmation": "",
      }
    }
    onSubmit={[Function]}
    validationSchema={
      ObjectSchema {
        "_blacklist": RefSet {
          "list": Set {},
          "refs": Map {},
        },
        "_conditions": Array [],
        "_defaultDefault": [Function],
        "_deps": Array [],
        "_excludedEdges": Array [],
        "_exclusive": Object {},
        "_mutate": undefined,
        "_nodes": Array [
          "passwordConfirmation",
          "password",
        ],
        "_options": Object {
          "abortEarly": true,
          "recursive": true,
        },
        "_type": "object",
        "_typeError": [Function],
        "_whitelist": RefSet {
          "list": Set {},
          "refs": Map {},
        },
        "fields": Object {
          "password": StringSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_deps": Array [],
            "_exclusive": Object {
              "required": true,
            },
            "_mutate": undefined,
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "string",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "tests": Array [
              [Function],
            ],
            "transforms": Array [
              [Function],
            ],
            "type": "string",
          },
          "passwordConfirmation": StringSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_deps": Array [],
            "_exclusive": Object {
              "required": true,
            },
            "_mutate": undefined,
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "string",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {
                null,
              },
              "refs": Map {
                "password" => Reference {
                  "getter": [Function],
                  "isContext": false,
                  "isSibling": true,
                  "isValue": false,
                  "key": "password",
                  "map": undefined,
                  "path": "password",
                },
              },
            },
            "_whitelistError": [Function],
            "tests": Array [
              [Function],
            ],
            "transforms": Array [
              [Function],
            ],
            "type": "string",
          },
        },
        "tests": Array [],
        "transforms": Array [
          [Function],
        ],
        "type": "object",
      }
    }
  >
    <Form
      className="password-form row"
    >
      <form
        action="#"
        className="password-form row"
        onReset={[Function]}
        onSubmit={[Function]}
      >
        <div
          className="mt-4 col-6"
        >
          <p>
            Password
          </p>
          <Field
            className="form-control"
            name="password"
            onChange={[Function]}
            placeholder="password"
            type="password"
          >
            <input
              className="form-control"
              name="password"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="password"
              type="password"
              value=""
            />
          </Field>
        </div>
        <div
          className=" mt-4  col-6 d-flex flex-column justify-content-center"
        >
          <div
            className="password-strength"
          >
            <p>
              Password strength
            </p>
            <div
              className="bar "
              data-testid="indicator-bar"
            />
          </div>
        </div>
        <div
          className="mt-4 col-6"
        >
          <p>
            Confirm Password
          </p>
          <Field
            className="form-control"
            name="passwordConfirmation"
            placeholder="password Confirmation"
            type="password"
          >
            <input
              className="form-control"
              name="passwordConfirmation"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="password Confirmation"
              type="password"
              value=""
            />
          </Field>
        </div>
        <Col
          xs={12}
        >
          <div
            className="col-12"
          >
            <Row
              noGutters={false}
            >
              <div
                className="row"
              >
                <Col
                  xs={6}
                >
                  <div
                    className="col-6"
                  >
                    <button
                      className="btn mt-3"
                      type="submit"
                    >
                      Save
                    </button>
                  </div>
                </Col>
              </div>
            </Row>
          </div>
        </Col>
      </form>
    </Form>
  </Formik>
</EditPassword>
`;
